test/features/push_notification/presentation/controller/push_base_controller_test.dart

```diff
@@ -0,0 +1,230 @@
+import 'package:jmap_dart_client/jmap/core/user_name.dart';
+import 'package:mockito/annotations.dart';
+import 'package:mockito/mockito.dart';
+import 'package:tmail_ui_user/features/push_notification/presentation/action/push_notification_state_change_action.dart';
+import 'package:tmail_ui_user/features/push_notification/presentation/controller/push_base_controller.dart';
+import 'package:tmail_ui_user/features/push_notification/presentation/listener/email_change_listener.dart';
+import 'package:tmail_ui_user/features/push_notification/presentation/listener/mailbox_change_listener.dart';
+
+import 'push_base_controller_test.mocks.dart';
+
+class TestPushController extends PushBaseController {
+  @override
+  void handleFailureViewState(Failure failure) {}
+
+  @override
+  void handleSuccessViewState(Success success) {}
+}
+
+@GenerateNiceMocks([
+  MockSpec<EmailChangeListener>(),
+  MockSpec<MailboxChangeListener>(),
+])
+void main() {
+  final accountId = AccountId(Id('accountId'));
+  final userName = UserName('userName');
+
+  group('push base controller test:', () {
+    group('mappingTypeStateToAction:', () {
+      final emailChangeListener = MockEmailChangeListener();
+      final mailboxChangeListener = MockMailboxChangeListener();
+
+      test(
+        'should call emailChangeListener.dispatchActions with SynchronizeEmailOnForegroundAction '
+        'when mapTypeState contains emailType '
+        'and isForeground is true',
+      () {
+        // arrange
+        final state = State('some-state');
+        final mapTypeState = {TypeName.emailType.value: state.value};
+
+        // act
+        final pushBaseController = TestPushController();
+        pushBaseController.mappingTypeStateToAction(
+          mapTypeState,
+          accountId,
+          userName,
+          isForeground: true,
+          emailChangeListener: emailChangeListener,
+          mailboxChangeListener: mailboxChangeListener
+        );
+
+        // assert
+        verify(
+          emailChangeListener.dispatchActions([
+            SynchronizeEmailOnForegroundAction(
+              TypeName.emailType,
+              state,
+              accountId,
+              null,
+            ),
+          ]),
+        ).called(1);
+      });
+
+      test(
+        'should call emailChangeListener.dispatchActions with StoreEmailStateToRefreshAction '
+        'when mapTypeState contains emailType '
+        'and isForeground is false',
+      () {
+        // arrange
+        final state = State('some-state');
+        final mapTypeState = {TypeName.emailType.value: state.value};
+
+        // act
+        final pushBaseController = TestPushController();
+        pushBaseController.mappingTypeStateToAction(
+          mapTypeState,
+          accountId,
+          userName,
+          isForeground: false,
+          emailChangeListener: emailChangeListener,
+          mailboxChangeListener: mailboxChangeListener
+        );
+
+        // assert
+        verify(
```

Comment:
should make sure `mailboxChangeListener` should not be invoked

Reply:
Done